Sources/MongoSwift/ConnectionPool.swift

```diff
@@ -67,38 +93,76 @@ internal class ConnectionPool {
         }
     }
 
-    /// Closes the pool, cleaning up underlying resources. This method blocks as it sends `endSessions` to the server.
-    internal func shutdown() {
-        switch self.state {
-        case let .open(pool):
-            mongoc_client_pool_destroy(pool)
-        case .closed:
-            return
+    /// Closes the pool, cleaning up underlying resources. **This method blocks until all connections are returned to
+    /// the pool.**
+    internal func close() throws {
+        try self.stateLock.withLock {
+            switch self.state {
+            case let .open(pool):
+                self.state = .closing(pool: pool)
+            case .closing, .closed:
+                throw Self.PoolClosedError
+            }
+        }
+
+        // continually loop and wait to get all connections back before destroying the pool. release the lock on each
+        // iteration to allow other methods to acquire the lock.
+        var done = false
+        while !done {
+            try self.stateLock.withLock {
+                if self.connCount == 0 {
+                    switch self.state {
+                    case .open, .closed:
+                        throw InternalError(message: "ConnectionPool in unexpected state \(self.state) during close()")
+                    case let .closing(pool):
+                        mongoc_client_pool_destroy(pool)
+                        self.state = .closed
+                    }
+                    done = true
+                }
+            }
         }
-        self.state = .closed
     }
 
     /// Checks out a connection. This connection will return itself to the pool when its reference count drops to 0.
     /// This method will block until a connection is available.
     internal func checkOut() throws -> Connection {
-        switch self.state {
-        case let .open(pool):
-            return Connection(clientHandle: mongoc_client_pool_pop(pool), pool: self)
-        case .closed:
-            throw InternalError(message: "ConnectionPool was already closed")
+        try self.stateLock.withLock {
+            switch self.state {
+            case let .open(pool):
+                self.connCount += 1
+                return Connection(clientHandle: mongoc_client_pool_pop(pool), pool: self)
+            case .closing, .closed:
+                throw Self.PoolClosedError
+            }
         }
     }
 
     /// Checks out a connection from the pool, or returns `nil` if none are currently available.
     internal func tryCheckOut() throws -> Connection? {
```

Comment:
nit: `try` implies a nil returning function but this also throws, might be worth adding that to the doc string. I know its in the declaration but I feel like its just slightly more future dev friendly.